ACM scoped policies

Charlesleonius opened the pull request:
* Adds support for Access Context Manager (ACM) Scoped Policies (Equivalent in [gcloud](https://cloud.google.com/sdk/gcloud/reference/access-context-manager/policies/create#--scopes))
* Adds support for IAM Policies for ACM Access Policies

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.


**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
access context manager: Added support for scoped policies in `google_access_context_manager_access_policy`
```

```release-note:new-resource
`google_access_context_manager_access_policy_iam_policy`
```

```release-note:new-resource
`google_access_context_manager_access_policy_iam_binding`
```

```release-note:new-resource
`google_access_context_manager_access_policy_iam_member`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@rileykarson, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 7 files changed, 405 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 7 files changed, 405 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 428 insertions(+), 129 deletions(-))

rileykarson commented on mmv1/products/accesscontextmanager/terraform.yaml:
```diff
@@ -26,11 +26,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         skip_test: true
         primary_resource_id: "access-policy"
+      - !ruby/object:Provider::Terraform::Examples
+        name: "access_context_manager_access_policy_scoped"
```
These are 1:1 with the template files so this won't quite work, just one of these is fine.

Charlesleonius replied:
Interesting, the documentation generated just fine. Two examples with the same name but different possible values. Is that not allowed?

Charlesleonius replied:
I've removed the duplicate just so that it won't cause issues when we eventually move the generated tests

rileykarson commented on mmv1/third_party/terraform/tests/resource_access_context_manager_access_policy_test.go.erb:
```diff
@@ -164,3 +165,34 @@ resource "google_access_context_manager_access_policy" "test-access" {
+resource "google_access_context_manager_access_policy" "test-access" {
+  parent = "organizations/%s"
+  title  = "%s"
+  scopes = ["projects/%s"]
```
Does the use of the scope field mean that there can now be multiple access policies per org?

Charlesleonius replied:
Yes, so long as the scopes are mutually exclusive

rileykarson replied:
Why would we want to wait, sorry?

rileykarson commented on mmv1/products/accesscontextmanager/terraform.yaml:
```diff
@@ -26,11 +26,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
     skip_sweeper: true
     id_format: "{{name}}"
     import_format: ["{{name}}"]
+    iam_policy: !ruby/object:Api::Resource::IamPolicy
```
Do the IAM & scope change need to happen simultaneously, or can they be broken up? Smaller, focused PRs are generally preferable & reduce the blasting radius of cherrypicks (by customers) and/or reverts (by us).

Charlesleonius replied:
As long as they can be included in the same release, I have no problem splitting them up.

rileykarson replied:
Barring a mixup they'll go out in the same release- if there's a strong need to ensure they're in _exactly_ the same one, keeping them together makes sense though.

Charlesleonius replied:
In that case, it would make sense to keep them together. They are very tightly coupled.

rileykarson reviewed (commented):
Couple questions, although I'd like to wait until the feature launch to run the tests (the time spent on access lists is typically more expensive than just running post-release)

rileykarson commented on mmv1/templates/terraform/examples/access_context_manager_access_policy_scoped.tf.erb:
```diff
@@ -0,0 +1,5 @@
+resource "google_access_context_manager_access_policy" "<%= ctx[:primary_resource_id] %>" {
+  parent = "organizations/123456789"
```
We'll need to pull the org of the test run from the environment rather than use a dummy value (and a dummy value will get injected into the docs). To do that, look for `test_env_vars` for an example- there are some semi-magic keys that will do so.

Charlesleonius replied:
Done

rileykarson commented on mmv1/templates/terraform/examples/access_context_manager_access_policy_scoped.tf.erb:
```diff
@@ -0,0 +1,5 @@
+resource "google_access_context_manager_access_policy" "<%= ctx[:primary_resource_id] %>" {
+  parent = "organizations/123456789"
+  title  = "my policy"
+  scopes = ["<%= ctx[:vars]['scope'] %>"]
```
Rather than plug in a scopes value, let's create a project in this configuration, reference it here (specifically using a Terraform reference i.e. `google_project.my_project.project_id` or equivalent for #, and then use it for the scoping project. That way, we'll ensure uniqueness across tests.

rileykarson commented:
I'm out for a few days, picking a new reviewer.

rileykarson commented:
Ah, guess it doesn't work because I'm assigned already & in the team. I'll pick @slevenick in that case!

Charlesleonius commented on mmv1/products/accesscontextmanager/terraform.yaml:
```diff
@@ -26,11 +26,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         primary_resource_id: "access-policy"
+      - !ruby/object:Provider::Terraform::Examples
+        name: "access_context_manager_access_policy_scoped"
+        skip_test: true
```
There is another blocker prevent us from using auto-generated tests at the moment. We can discuss this out of band.

Charlesleonius commented:
I tried to migrate tests from manual to auto-generated (using examples) but ran into some issues due to IAM tests not supporting the import of resources with a server side generated id. These changes are somewhat time sensitive so it would be preferable to put off changing the tests till a later date. Can you please try running the integration tests to make sure everything works as is.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 7 files changed, 401 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 7 files changed, 401 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 429 insertions(+), 130 deletions(-))

slevenick commented:
/gcbrun

Have you been able to run tests for the IAM policy? If the generated tests do not work can we handwrite some tests to make sure this is all tested?

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 7 files changed, 401 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 7 files changed, 401 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 429 insertions(+), 130 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 525 insertions(+), 3 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 525 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 429 insertions(+), 130 deletions(-))

slevenick commented:
Looks like the lint is failing due to:

google-beta/resource_access_context_manager_access_policy_iam_test.go:70:6: `testAccAccessContextManagerAccessPolicyIamMember` is unused (deadcode)
func testAccAccessContextManagerAccessPolicyIamMember(policy, account, role string) string {
     ^
GNUmakefile:27: recipe for target 'lint' failed

Charlesleonius commented:
> Looks like the lint is failing due to:
> 
> google-beta/resource_access_context_manager_access_policy_iam_test.go:70:6: `testAccAccessContextManagerAccessPolicyIamMember` is unused (deadcode) func testAccAccessContextManagerAccessPolicyIamMember(policy, account, role string) string { ^ GNUmakefile:27: recipe for target 'lint' failed

Added back the test that was using this method. It was having issues previously so I removed it.

Charlesleonius commented:
@slevenick Please run vcr tests again when you can. I added new tests to cover the iam resources.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 429 insertions(+), 130 deletions(-))

rileykarson commented:
/gcbrun

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 432 insertions(+), 133 deletions(-))

modular-magician commented:
Tests were added that did not run in TeamCity:

- TestAccAccessContextManagerAccessPolicyIamBinding
- TestAccAccessContextManagerAccessPolicyIamMember
- TestAccAccessContextManagerAccessPolicyIamPolicy

modular-magician commented:
I have triggered VCR tests in RECORDING mode for the following tests that failed during VCR: TestAccDataprocMetastoreServiceDatasource_basic|TestAccAccessContextManager|TestAccContainerCluster_withAuthenticatorGroupsConfig|TestAccSpannerDatabase_postgres You can view the result here: https://ci-oss.hashicorp.engineering/viewQueued.html?itemId=274389

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 432 insertions(+), 133 deletions(-))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-5785-old..auto-pr-5785) ( 8 files changed, 549 insertions(+), 3 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-5785-old..auto-pr-5785) ( 6 files changed, 432 insertions(+), 133 deletions(-))

rileykarson commented:
Just helping to unstick the tests!

slevenick commented:
/gcbrun

slevenick reviewed (commented):
Are you able to run the tests?

I'm getting a 500 error attempting to run it:
`Error setting IAM policy for accesscontextmanager accesspolicy "accessPolicies/xxxx": googleapi: Error 500: Internal error encountered.`

Charlesleonius commented:
> Are you able to run the tests?
> 
> I'm getting a 500 error attempting to run it: `Error setting IAM policy for accesscontextmanager accesspolicy "accessPolicies/xxxx": googleapi: Error 500: Internal error encountered.`

@slevenick Hmm I just tried running the tests and they worked

slevenick commented:
Cool, seems like the tests pass most of the time (though the 500 error is scary) so this should be good to go @rileykarson ?

rileykarson commented:
Are you still getting the error if you try to run them @slevenick? We may need to get the service to expose more info or run from an internal IP for better logs. AccessPolicy related stuff has a habit of failing the whole accesscontextmanager suite, so it's probably best to play these safe since that's 4-5 resources.

slevenick commented:
> Are you still getting the error if you try to run them @slevenick? We may need to get the service to expose more info or run from an internal IP for better logs. AccessPolicy related stuff has a habit of failing the whole accesscontextmanager suite, so it's probably best to play these safe since that's 4-5 resources.

No, I got it twice in a row attempting to run the tests initially, but now it seems to be cleared up? I ran it twice successfully so I'm guessing the service was down

Charlesleonius commented:
> Are you still getting the error if you try to run them @slevenick? We may need to get the service to expose more info or run from an internal IP for better logs. AccessPolicy related stuff has a habit of failing the whole accesscontextmanager suite, so it's probably best to play these safe since that's 4-5 resources.

The access context manager test suite is very fragile right now. Next PR I will do some work to migrate it to the examples framework and see if I can make the tests a bit more resilient.
